File: Ballast.Client/src/messaging/ievent-bus.ts
import { IDisposable } from '../interfaces/idisposable';
import { IEvent } from './ievent';

export type AsyncEventHandler<TEvent extends IEvent> = (event: TEvent) => Promise<void>

export interface IEventBus extends IDisposable {
    getHandlers<TEvent extends IEvent>(type: Symbol): AsyncEventHandler<TEvent>[];
    publishAsync<TEvent extends IEvent>(event: TEvent): Promise<void>;
    subscribe<TEvent extends IEvent>(type: Symbol, handler: AsyncEventHandler<TEvent>): void;
    unsubscribe<TEvent extends IEvent>(type: Symbol, handler: AsyncEventHandler<TEvent>): void;
}

File: ballast-core/dotnet/src/messaging/IEventBus.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Ballast.Core.Messaging
{
    public interface IEventBus
    {
        IEnumerable<Func<TEvent, Task>> GetHandlers<TEvent>(string type) where TEvent : IEvent;
        Task PublishAsync<TEvent>(TEvent evt) where TEvent : IEvent;
        void Subscribe<TEvent>(string type, Func<TEvent, Task> asyncEventHandler) where TEvent : IEvent;
        void Unsubscribe<TEvent>(string type, Func<TEvent, Task> asyncEventHandler) where TEvent : IEvent;
    }
}